### PR TITLE
Better input handling, mouse cursor, completing the action obs loop

### PR DESCRIPTION
**Pull request by rk1a**
Python side
* add basic render function
* receive image observations
* add middle mouse button

C++ side
* send image observations with zmq socket of input handler
* fix GUIEvents not being triggered
* mouse buttons specific handling
* menu/inventory specific handling
* overlay mouse cursor when GUI is open
* add option to prevent screen resizing

## To do

This PR is a Work in Progress

- [x] Use protobuf for data (action/observation) serialization
- [ ] Use better zmq patterns than request-reply

## How to test

Start `server.sh`, `client.sh` and `test_loop.py`

### CODE REVIEW

File: hacking_testing/minetest_env.py [hunk 1]
```diff
@@ -1,6 +1,19 @@
 import gym
+import matplotlib.pyplot as plt
+import numpy as np
 import zmq
 from gym.spaces import Box, Dict, Discrete
+from matplotlib.pyplot import close, draw, figure, pause
+
+plt.rcParams["toolbar"] = "None"
```
Comment (neverix) [inline hacking_testing/minetest_env.py]: Move into test_loop or rendering functions?

File: src/client/game.cpp [hunk 2]
```diff
@@ -1172,10 +1201,18 @@ void Game::run()
 
 	irr::core::dimension2d<u32> previous_screen_size(g_settings->getU16("screen_w"),
 		g_settings->getU16("screen_h"));
-
+
```
Comment (neverix) [inline src/client/game.cpp]: Spacing bug

File: src/client/game.cpp [hunk 3]
```diff
@@ -1139,8 +1148,28 @@ bool Game::startup(bool *kill,
 	if (!createClient(start_data))
 		return false;
 
-	if(start_data.isRecording())
+	// create ZMQ objects
+	std::string address = std::string("tcp://127.0.0.1:") + start_data.dumb_client_port;
```
Comment (neverix) [inline src/client/game.cpp:1152]: What if we want to connect to a remote machine? Possible solution: host a router on the host PC.
Also, this removes the need for `dumb-port`. If this change goes through, the two arguments have to be renamed and merged.

**Comment by neverix** [comment]
I switched to a protobuf backend in [my fork](https://github.com/neverix/minetest/tree/input-recorder-improvements), can you pull from there?

**Comment by neverix** [comment]
Also, `step()` should be in `dumbhandler.cpp` to follow the convention for the random input handler and `inputhandler.cpp`

File: src/client/game.cpp (hunk 3 above)
Comment (rk1a) [reply src/client/game.cpp:1152]: Yes, good point. I mean we can still have a configurable address, maybe just rename `dumb-port` to `dumb-address`?

File: .gitignore [hunk 4]
```diff
@@ -37,6 +37,14 @@ build/.cmake/
 # GNU Patch reject file
 *.rej
 
+# Protobuf
+dumb_inputs.pb.cc
```
Comment (neverix) [inline .gitignore]: Replace with `**/*.pb.cc`, `**/*.pb.h` and `**/*_pb2.py`?

**Pull request merged**